src/blocks/core/group.js

```diff
@@ -0,0 +1,203 @@
+const withCustomGroupControls = createHigherOrderComponent( ( BlockEdit ) => {
+	return ( props ) => {
+		if ( ! props.isSelected || props.name !== 'core/group' ) {
+			return <BlockEdit key="edit" { ...props } />;
+		}
+		const { attributes, setAttributes } = props;
+
+		return (
+			<>
+				<BlockEdit key="edit" { ...props } />
+				<InspectorControls>
+					<PanelBody
+						title={ __( 'Report group settings', 'wmf-reports' ) }
+						initialOpen={ true }
+					>
+						<ToggleControl
+							label={ __(
+								'Include in Table of Contents',
+								'wmf-reports'
+							) }
+							checked={ attributes.includeInToC }
+							onChange={ ( includeInToC ) => {
+								setAttributes( { includeInToC } );
+							} }
+						/>
+						{ attributes.includeInToC && (
+							<TextControl
+								label={ __(
+									'Table of Contents Label',
+									'wmf-reports'
+								) }
+								value={ attributes.tocLabel }
+								onChange={ ( tocLabel ) => {
+									const id = toIdString( tocLabel );
+									setAttributes( {
+										tocSlug: `toc-${ id }`,
+										tocLabel,
+									} );
+								} }
+							/>
+						) }
+						<ToggleControl
+							label={ __(
+								'Fill viewport height',
+								'wmf-reports'
+							) }
+							checked={ attributes.fullViewportHeight }
+							onChange={ ( fullViewportHeight ) => {
+								setAttributes( { fullViewportHeight } );
+							} }
+							help={ __(
+								'Makes this group take up at minimum a full viewport of vertical space',
+								'wmf-reports'
+							) }
+						/>
+					</PanelBody>
+				</InspectorControls>
+			</>
+		);
+	};
+}, 'withCustomGroupControls' );
+
+addFilter(
+	'editor.BlockEdit',
+	'wmf-reports/add-group-block-controls',
+	withCustomGroupControls
+);
+
+/**
+ * Allow the "fill viewport height" toggle to be previewed. The style override
+ * is added on the frontend using a PHP filter, but we use JS to filter in-editor
+ * to add the override classname.
+ */
+const withFillViewportHeightStyle = createHigherOrderComponent(
+	( BlockListBlock ) => {
+		return ( props ) => {
+			if (
+				props.name !== 'core/group' ||
+				! props.attributes.fullViewportHeight
+			) {
+				return <BlockListBlock { ...props } />;
+			}
+			const mergedCssClasses = `${
+				props.className || ''
+			} wp-block-group-is-full-viewport-height`;
+			return (
+				<BlockListBlock { ...props } className={ mergedCssClasses } />
+			);
+		};
+	},
+	'withFillViewportHeightStyle'
+);
+
+addFilter(
+	'editor.BlockListBlock',
+	'wmf-reports/block-list-block/group-block',
+	withFillViewportHeightStyle,
+	100
+);
+
+/**
+ * Add additional custom properties to group blocks.
+ *
+ * @todo Should this be restricted to groups on Report post type objects?
+ *
+ * @param {Object} settings Block settings object.
+ * @param {string} name     Block type name.
+ * @return {Object} Filtered settings object.
+ */
+function customizeGroupBlockAttributes( settings, name ) {
+	if ( name !== 'core/group' ) {
```

Comment:
Do we need other blocks to support the TOC attribute? If so, this function will have to be refactored in the future. But no blocker now, and it's always possible to wrap any other block in a group block.